staging/src/k8s.io/legacy-cloud-providers/azure/azure.go

```diff
@@ -495,28 +511,45 @@ func (az *Cloud) InitializeCloudFromConfig(config *Config, fromSecret bool) erro
+
 	az.InterfacesClient = interfaceclient.New(azClientConfig.WithRateLimiter(config.InterfaceRateLimit))
-	az.RouteTablesClient = routetableclient.New(azClientConfig.WithRateLimiter(config.RouteTableRateLimit))
-	az.LoadBalancerClient = loadbalancerclient.New(azClientConfig.WithRateLimiter(config.LoadBalancerRateLimit))
-	az.SecurityGroupsClient = securitygroupclient.New(azClientConfig.WithRateLimiter(config.SecurityGroupRateLimit))
-	az.PublicIPAddressesClient = publicipclient.New(azClientConfig.WithRateLimiter(config.PublicIPAddressRateLimit))
-	az.VirtualMachineScaleSetsClient = vmssclient.New(azClientConfig.WithRateLimiter(config.VirtualMachineScaleSetRateLimit))
 	az.VirtualMachineSizesClient = vmsizeclient.New(azClientConfig.WithRateLimiter(config.VirtualMachineSizeRateLimit))
 	az.SnapshotsClient = snapshotclient.New(azClientConfig.WithRateLimiter(config.SnapshotRateLimit))
 	az.StorageAccountClient = storageaccountclient.New(azClientConfig.WithRateLimiter(config.StorageAccountRateLimit))
-	az.VirtualMachinesClient = vmclient.New(azClientConfig.WithRateLimiter(config.VirtualMachineRateLimit))
 	az.DisksClient = diskclient.New(azClientConfig.WithRateLimiter(config.DiskRateLimit))
 	// fileClient is not based on armclient, but it's still backoff retried.
 	az.FileClient = newAzureFileClient(env, azClientConfig.Backoff)
 
+	vmClientConfig := azClientConfig.WithRateLimiter(config.VirtualMachineRateLimit)
+	vmClientConfig.MultiTenantServicePrincipalToken = multiTenantServicePrincipalToken
+	az.VirtualMachinesClient = vmclient.New(vmClientConfig)
+
+	vmssClientConfig := azClientConfig.WithRateLimiter(config.VirtualMachineScaleSetRateLimit)
+	vmssClientConfig.MultiTenantServicePrincipalToken = multiTenantServicePrincipalToken
+	az.VirtualMachineScaleSetsClient = vmssclient.New(vmssClientConfig)
+
 	// Error "not an active Virtual Machine Scale Set VM" is not retriable for VMSS VM.
 	// But http.StatusNotFound is retriable because of ARM replication latency.
 	vmssVMClientConfig := azClientConfig.WithRateLimiter(config.VirtualMachineScaleSetRateLimit)
 	vmssVMClientConfig.Backoff = vmssVMClientConfig.Backoff.WithNonRetriableErrors([]string{vmssVMNotActiveErrorMessage}).WithRetriableHTTPStatusCodes([]int{http.StatusNotFound})
+	vmssVMClientConfig.MultiTenantServicePrincipalToken = multiTenantServicePrincipalToken
 	az.VirtualMachineScaleSetVMsClient = vmssvmclient.New(vmssVMClientConfig)
 
+	networkResourceClientConfig := azClientConfig
+	if networkResourceServicePrincipalToken != nil {
+		newClientConfig := *networkResourceClientConfig
+		newClientConfig.SubscriptionID = config.CrossTenantNetworkResourceConfig.SubscriptionID
+		newClientConfig.ServicePrincipalToken = networkResourceServicePrincipalToken
+		networkResourceClientConfig = &newClientConfig
+	}
+
+	az.RoutesClient = routeclient.New(networkResourceClientConfig.WithRateLimiter(config.RouteRateLimit))
```

Comment:
seems this would break none-multi-tenant cases? as servicePrincipalToken is not used any more

Reply:
No actually. For non-multi-tenant cases, networkResourceServicePrincipalToken will be null so networkResourceClientConfig will be azClientConfig, which is the same as original logic without this new feature.

Reply:
The confusion might caused by this my unclear code structure. Please take a look at the refactored code. Thank you.

Reply:
@feiskyer I keep this conversation open. Please take a look and make sure I didn't miss/break anything. Thank you.

Reply:
what about write `configAzureClients` as following:
```
configAzureClients(servicePrincipalToken, multiTenantServicePrincipalToken, networkResourceServicePrincipalToken)
```
`multiTenantServicePrincipalToken`, `networkResourceServicePrincipalToken` could be nil, and outside of `configAzureClients`, handling logic of 
```
	if az.Config.UsesNetworkResourceInDifferentTenant() {
		multiTenantServicePrincipalToken, err := auth.GetMultiTenantServicePrincipalToken(&az.Config.AzureAuthConfig, &az.Environment)
		if err != nil {
			return err
		}
		networkResourceServicePrincipalToken, err := auth.GetNetworkResourceServicePrincipalToken(&az.Config.AzureAuthConfig, &az.Environment)
		if err != nil {
			return err
		}
    ...
```

Reply:
Sure. Updated. Thank you.